docs/ALU Design.md

```diff
@@ -0,0 +1,758 @@
+```
+
+Unlike the previous example, this snippet does nto contain the `add` instruction. The adding
+operation is done by an intrinsic named `llvm.uadd.with.overflow.i64`, which is called with the
+`call` instruction. The intrinsic is defined in the LLVM codebase and its source code does not make
+it into the `.ll` file produced out of the adding operation in Rust code.
+
+The LLVM Language Reference Manual has an extensive list of intrinsics. Here's the example of
+[`llvm.uadd.with.overflow.<ty>`](https://llvm.org/docs/LangRef.html#llvm-uadd-with-overflow-intrinsics).
+
+### Data types
+
+This ALU will support only integers. Floating point numbers are out of scope of this document.
+
+#### Integers
+
+LLVM IR supports integers of arbitrary width. A general syntax for an integer type is `iN`, where
+`N` can be anything from 1 to 2^32. Since LLVM does not have a dedicated type for boolean values,
+`i1` is used instead.
+
+The Cairo VM internally operates on 252-bit-long field elements - `felt252`. On the higher level of
+abstraction, the Cairo language supports
+[integers of specific lengths](https://book.cairo-lang.org/ch02-02-data-types.html): 8 bit, 16 bit,
+32 bit, 64 bit, 128 bit and 256 bit. Cairo also supports booleans.
+
+[Rust supports integers of width from 8 to 128 bit](https://doc.rust-lang.org/book/ch03-02-data-types.html)
+with the same increment Cairo does, plus architecture-dependent `isize` and `usize`. Rust also
+supports booleans.
+
+The Cairo VM does not have a classical registers of length constrained by the hardware. Therefore
+there is no obvious indicator of how long `usize`/`isize` should be on that target. Since from the
+LLVM point of view a pointer must have a finite size, this decision must be made based on some other
+feature of the architecture. We have evaluated the following choices:
+
+- The Cairo language already has 32 bit `usize`, so we can follow this approach, making `usize` and
+  `isize` also 32 bit,
+- The architecture's natural word size is 252 bit, being the length of the field element. It may be
+  reasonable to set `usize` and `isize` length to 252 bit,
+- 256 bit is the next power-of-2 after 252. Having `usize` and `isize` 256 bit long leaves 4 extra
+  bits that may be used to keep some metadata.
+
+Ultimately the size of `usize` and `isize` has been decided to be 64 bits, which is neither of the
+above possibilities. This length is a consequence of using the `aarch64-unknown-none-softfloat`
+target triple. The choice of the triple determines the length of the pointer which in turn
+determines the length of `usize`. This target triple is a temporary choice before a custom target
+triple is proposed. It has been chosen for its soft float support and no host operating system. The
+pointer length is just one of its parameters we accept on this stage of the project.
+
+Summing up, we expect to see in the IR integers of the following lengths: 1, 8, 16, 32, 64 and 128
```

Comment:
Which IR? If you mean LLVM IR we are unlikely to see `i128`, at least to start with.

Reply:
At least initially we can construct our Rust programs to avoid that type! Furthermore, our target (`aarch64-unknown-none-softfloat`) does not have native support for them at the moment, so we are safe to bail _for now_.